sdks/java/io/tika/src/main/java/org/apache/beam/sdk/io/tika/TikaSource.java

```diff
@@ -0,0 +1,466 @@
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.NoSuchElementException;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import javax.annotation.Nullable;
+
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.StringUtf8Coder;
+import org.apache.beam.sdk.io.BoundedSource;
+import org.apache.beam.sdk.io.FileSystems;
+import org.apache.beam.sdk.io.Source;
+import org.apache.beam.sdk.io.fs.MatchResult;
+import org.apache.beam.sdk.io.fs.MatchResult.Metadata;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.tika.config.TikaConfig;
+import org.apache.tika.exception.TikaException;
+import org.apache.tika.io.TikaInputStream;
+import org.apache.tika.parser.AutoDetectParser;
+import org.apache.tika.parser.ParseContext;
+import org.apache.tika.parser.Parser;
+import org.joda.time.Instant;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.xml.sax.SAXException;
+import org.xml.sax.helpers.DefaultHandler;
+
+/**
+ * Implementation detail of {@link TikaIO.Read}.
+ *
+ * <p>A {@link Source} which can represent the content of the files parsed by Apache Tika.
+ */
+class TikaSource extends BoundedSource<String> {
+  private static final long serialVersionUID = -509574062910491122L;
+  private static final Logger LOG = LoggerFactory.getLogger(TikaSource.class);
+
+  @Nullable
+  private MatchResult.Metadata singleFileMetadata;
+  private final Mode mode;
+  private final TikaIO.Read spec;
+
+  /**
+   * Source mode.
+   */
+  public enum Mode {
+    FILEPATTERN, SINGLE_FILE
+  }
+
+  TikaSource(TikaIO.Read spec) {
+    this.mode = Mode.FILEPATTERN;
+    this.spec = spec;
+  }
+
+  TikaSource(Metadata fileMetadata, TikaIO.Read spec) {
+    mode = Mode.SINGLE_FILE;
+    this.singleFileMetadata = checkNotNull(fileMetadata, "fileMetadata");
+    this.spec = spec;
+  }
+
+  @Override
+  public BoundedReader<String> createReader(PipelineOptions options) throws IOException {
+    this.validate();
+    checkState(spec.getFilepattern().isAccessible(),
+        "Cannot create a Tika reader without access to the file"
+        + " or pattern specification: {}.", spec.getFilepattern());
+    if (spec.getTikaConfigPath() != null) {
+      checkState(spec.getTikaConfigPath().isAccessible(),
+        "Cannot create a Tika reader without access to its configuration",
+        spec.getTikaConfigPath());
+    }
+
+    String fileOrPattern = spec.getFilepattern().get();
```

Comment:
Copy-pasting code from FileBasedSource creates a maintenance debt - this code will not receive bugfixes or other improvements to FileBasedSource. If FileBasedSource lacked a particular feature, it is entirely reasonable to propose to make FileBasedSource have the feature, rather than copy-paste its code: copy-paste of a "not quite right" library is something acceptable in a user's code, but not in the library itself.

Reply:
I agree; at a time I was creating a PR I was not quite comfortable with source and I also wanted to make the initial PR self-contained - I'll look at at dropping this code after I address other issues you pointed to...